### PR TITLE
Aggregate alerts

### PR DESCRIPTION
**Pull request by fjerlov-cs**
Adds aggregate-alerts to humio/cli

```
➜  cli git:(fjerlov/aggregate-alert) ✗ ./bin/humioctl aggregate-alerts
Manage aggregate alerts

Usage:
  humioctl aggregate-alerts [command]

Available Commands:
  export      Export a aggregate alert <aggregate-alert> in <view> to a file.
  install     Installs a aggregate alert in a view
  list        List all aggregate alerts in a view.
  remove      Removes a aggregate alert.
  show        Show details about a aggregate alert in a view.
```

```
/Users/fjerlov/Library/Caches/JetBrains/GoLand2023.3/tmp/GoLand/___list aggregate-alerts list DemoView
                 ID                |      NAME       |  DESCRIPTION  | ACTION NAMES | LABELS | ENABLED | THROTTLE FIELD | THROTTLE TIME SECONDS | SEARCH INTERVAL SECONDS | QUERY TIMESTAMP TYPE | TRIGGER MODE  |      RUN AS USERID       | QUERY OWNERSHIP TYPE  
+----------------------------------+-----------------+---------------+--------------+--------+---------+----------------+-----------------------+-------------------------+----------------------+---------------+--------------------------+----------------------+
  SvhhCj21ZV9dRkQd0FBohuCeUVQly77a | aggregate-alert | asdakmsdasd-2 | testing      | label2 | false   | asd            |                  7200 |                    7200 | EventTimestamp       | ImmediateMode | pMeYeiIlF3NljnMAIcEcRnUK | User                  
```

```
/Users/fjerlov/Library/Caches/JetBrains/GoLand2023.3/tmp/GoLand/___show aggregate-alerts show DemoView aggregate-alert
                       ID | SvhhCj21ZV9dRkQd0FBohuCeUVQly77a  
                     Name | aggregate-alert                   
              Description | asdakmsdasd-2                     
             Query String | foo=* bar=* | count()             
  Search Interval Seconds | 7200                              
                  Actions | testing                           
                   Labels | label2                            
                  Enabled | false                             
           Throttle Field | asd                               
    Throttle Time Seconds | 7200                              
     Query Timestamp Type | EventTimestamp                    
             Trigger Mode | ImmediateMode                     
           Run As User ID | pMeYeiIlF3NljnMAIcEcRnUK          
     Query Ownership Type | User                              

Process finished with the exit code 0

```

aggregate-alert.yaml
```
name: aggregate-alert
description: asdakmsdasd-2
queryString: foo=* bar=* | count()
searchIntervalSeconds: 7200
actionNames:
- testing
labels:
- label2
enabled: false
throttleField: asd
throttleTimeSeconds: 7200
queryOwnershipType: User
triggerMode: ImmediateMode
queryTimestampType: EventTimestamp
runAsUserId: pMeYeiIlF3NljnMAIcEcRnUK
```

### CODE REVIEW

**Review by jespermbyskov — APPROVED** [review]
LGTM

File: api/internal/humiographql/aggregate-alerts.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package humiographql
+
+import graphql "github.com/cli/shurcooL-graphql"
+
+type AggregateAlert struct {
+	ID                    graphql.String     `graphql:"id"`
```
Comment (fjerlov-cs) [inline api/internal/humiographql/aggregate-alerts.go:6]: should be json not graphql - copy paste error

File: api/internal/humiographql/aggregate-alerts.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	Name                  graphql.String     `graphql:"name"`
```
Comment (fjerlov-cs) [inline api/internal/humiographql/aggregate-alerts.go:7]: should be json not graphql - copy paste error

**Review by SaaldjorMike — APPROVED** [review]

File: cmd/humioctl/aggregate_alerts_export.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+)
+
+func newAggregateAlertsExportCmd() *cobra.Command {
+	var outputName string
+
+	cmd := cobra.Command{
+		Use:   "export [flags] <view> <aggregate-alert>",
```
Comment (danielamkaer) [inline cmd/humioctl/aggregate_alerts_export.go:29]: is `<aggregate-alert>` a name or ID?

File: cmd/humioctl/aggregate_alerts_export.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			view := args[0]
+			aggregateAlertName := args[1]
+			client := NewApiClient(cmd)
+
+			if outputName == "" {
+				outputName = aggregateAlertName
+			}
+
+			aggregateAlerts, err := client.AggregateAlerts().List(view)
```
Comment (danielamkaer) [inline cmd/humioctl/aggregate_alerts_export.go:41]: Why does this not use the `Get` method?

File: cmd/humioctl/aggregate_alerts_show.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	cmd := cobra.Command{
+		Use:   "show <view> <name>",
+		Short: "Show details about an aggregate alert in a view.",
+		Args:  cobra.ExactArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			view := args[0]
+			name := args[1]
+			client := NewApiClient(cmd)
+
+			aggregateAlerts, err := client.AggregateAlerts().List(view)
```
Comment (danielamkaer) [inline cmd/humioctl/aggregate_alerts_show.go:34]: Why not `Get`?

**Review by danielamkaer — APPROVED** [review]

File: cmd/humioctl/aggregate_alerts_show.go (hunk 5 above)
Comment (fjerlov-cs) [reply cmd/humioctl/aggregate_alerts_show.go:34]: We don't have use the Id in the cli.

File: cmd/humioctl/aggregate_alerts_export.go (hunk 4 above)
Comment (fjerlov-cs) [reply cmd/humioctl/aggregate_alerts_export.go:41]: We don't use the Id in the cli. Get takes Id.

File: cmd/humioctl/aggregate_alerts_export.go (hunk 3 above)
Comment (fjerlov-cs) [reply cmd/humioctl/aggregate_alerts_export.go:29]: Name, by convention.

File: api/aggregate-alerts.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+		return nil, err
+	}
+
+	aggregateAlert := mapHumioGraphqlAggregateAlertToAggregateAlert(mutation.AggregateAlert)
+
+	return &aggregateAlert, nil
+}
+
+func (a *AggregateAlerts) Delete(viewName, aggregateAlertID string) error {
+	if aggregateAlertID == "" {
```
[the remark this answers is not in the document]
Comment (fjerlov-cs) [reply api/aggregate-alerts.go:162]: I believe it is an oversight, as the viewName gets checked by the cli, but not for Go API use. I'll add the checks.

File: api/aggregate-alerts.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+	RunAsUserID           string   `graphql:"runAsUserId"           yaml:"runAsUserId,omitempty"   json:"runAsUserId,omitempty"`
+}
+
+type AggregateAlerts struct {
+	client *Client
+}
+
+func (c *Client) AggregateAlerts() *AggregateAlerts { return &AggregateAlerts{client: c} }
+
+func (a *AggregateAlerts) List(viewName string) ([]AggregateAlert, error) {
```
[the remark this answers is not in the document]
Comment (fjerlov-cs) [reply api/aggregate-alerts.go:32]: Fixed

File: api/aggregate-alerts.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	var aggregateAlerts = make([]AggregateAlert, len(query.SearchDomain.AggregateAlerts))
+	for i := range query.SearchDomain.AggregateAlerts {
+		aggregateAlerts[i] = mapHumioGraphqlAggregateAlertToAggregateAlert(query.SearchDomain.AggregateAlerts[i])
+	}
+
+	return aggregateAlerts, err
```
[the remark this answers is not in the document]
Comment (fjerlov-cs) [reply api/aggregate-alerts.go:53]: Nice find

**Comment by fjerlov-cs** [comment]
@danielamkaer with the new changes:

show:
```
/Users/fjerlov/Library/Caches/JetBrains/GoLand2024.1/tmp/GoLand/___show aggregate-alerts show DemoView aggregate-alert
                       ID | hooUSUSV36H3TP9VXJzi69ix55EaoYGQ  
                     Name | aggregate-alert                   
              Description | test 2 3                          
             Query String | foo=* bar=2 | count()             
  Search Interval Seconds | 3600                              
                  Actions | testaction, testing               
                   Labels | label1, label2, label3            
                  Enabled | true                              
           Throttle Field | test                              
    Throttle Time Seconds | 3600                              
     Query Timestamp Type | IngestTimestamp                   
             Trigger Mode | ImmediateMode                     
           Run As User ID |                                   
     Query Ownership Type | Organization     
```

export:
```
name: aggregate-alert
description: test 2 3
queryString: foo=* bar=2 | count()
searchIntervalSeconds: 3600
actionNames:
- testaction
- testing
labels:
- label1
- label2
- label3
enabled: true
throttleField: test
throttleTimeSeconds: 3600
queryOwnershipType: Organization
triggerMode: ImmediateMode
queryTimestampType: IngestTimestamp
```

List:
```
/Users/fjerlov/Library/Caches/JetBrains/GoLand2024.1/tmp/GoLand/___list aggregate-alerts list DemoView
                 ID                |      NAME       | DESCRIPTION |    ACTION NAMES     |         LABELS         | ENABLED | THROTTLE FIELD | THROTTLE TIME SECONDS | SEARCH INTERVAL SECONDS | QUERY TIMESTAMP TYPE | TRIGGER MODE  | RUN AS USERID | QUERY OWNERSHIP TYPE  
+----------------------------------+-----------------+-------------+---------------------+------------------------+---------+----------------+-----------------------+-------------------------+----------------------+---------------+---------------+----------------------+
  hooUSUSV36H3TP9VXJzi69ix55EaoYGQ | aggregate-alert | test 2 3    | testaction, testing | label1, label2, label3 | true    | test           |                  3600 |                    3600 | IngestTimestamp      | ImmediateMode |               | Organization          


```

**Pull request merged**